authentik/core/models.py

```diff
@@ -887,3 +897,15 @@ def from_request(request: HttpRequest, user: User) -> Optional["AuthenticatedSes
             last_user_agent=request.META.get("HTTP_USER_AGENT", ""),
             expires=request.session.get_expiry_date(),
         )
+
+
+class UserPasswordHistory(models.Model):
+    user = models.ForeignKey("User", on_delete=models.CASCADE, related_name="old_passwords")
+    change = models.JSONField()
```

Comment:
I opted for a JSON field over an individual `old_password` field for two reasons:

1. We'll likely never need to query on `old_password` and it's a good candidate for denormalization
2. It allows us to add additional additional data in this audit table without a migration. 

If Authentik prefers to avoid JSON columns I'm happy to switch this out.

Reply:
> Adding migrations is not that much of an issue for existing installs (especially adding a new column should be quite fast).

Wasn't sure if there were concerns around adding migrations or not. Sounds like it's not a problem, though. 

I've run into situations in the past where doing a migration would take a long time if the affected tables had an FK to another table that was frequently taking locks for `UPDATE` 

---
💥 I will change this model to use a field for the previous password instead of JSON and make updates accordingly.